File: leetcode/9.cpp
// Runtime: 42 ms, faster than 8.29% of C++ online submissions for Palindrome Number.
// Memory Usage: 5.8 MB, less than 74.02% of C++ online submissions for Palindrome Number.
class Solution {
 public:
  bool isPalindrome(int x) {
    // early exit
    if (x < 0) {
      return false;
    }

    // construct the number in reverse
    long res = 0;
    int x_copy = x;
    while (x) {
      res += x % 10;
      res *= 10;
      x /= 10;
    }

    return res / 10 == x_copy;
  }
};

File: leetcode/9.py
"""
Runtime: 114 ms, faster than 5.12% of Python3 online submissions for Palindrome Number.
Memory Usage: 14 MB, less than 99.04% of Python3 online submissions for Palindrome Number.
"""


class Solution:
    def isPalindrome(self, x: int) -> bool:
        if x < 0:
            return False
        x_pop_backward = x

        digits = []
        while x_pop_backward:
            x_pop_backward, r = divmod(x_pop_backward, 10)
            digits.append(r)

        i, j = 0, len(digits) - 1
        while i < j:
            if digits[i] != digits[j]:
                return False
            i += 1
            j -= 1

        return True


"""
Runtime: 106 ms, faster than 6.10% of Python3 online submissions for Palindrome Number.
Memory Usage: 14.1 MB, less than 92.66% of Python3 online submissions for Palindrome Number.
"""


class Solution:
    def isPalindrome(self, x: int) -> bool:
        return list(str(x)) == list(str(x))[::-1]
